src/main/java/org/medicmobile/webapp/mobile/ChtExternalAppLauncher.java

```diff
@@ -0,0 +1,388 @@
+package org.medicmobile.webapp.mobile;
+
+import static org.medicmobile.webapp.mobile.JavascriptUtils.safeFormat;
+import static org.medicmobile.webapp.mobile.MedicLog.error;
+import static org.medicmobile.webapp.mobile.MedicLog.trace;
+import static org.medicmobile.webapp.mobile.Utils.getUriFromFilePath;
+
+import android.app.Activity;
+import android.content.Intent;
+import android.graphics.Bitmap;
+import android.graphics.BitmapFactory;
+import android.net.Uri;
+import android.os.Bundle;
+import android.os.ParcelFileDescriptor;
+import android.util.Base64;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.ByteArrayOutputStream;
+import java.io.FileInputStream;
+import java.io.InputStream;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collector;
+
+public class ChtExternalAppLauncher {
+
+	private ChtExternalAppLauncher() { }
+
+	static Intent createIntent(ChtExternalApp chtExternalApp) {
+		ChtExternalAppIntentBuilder builder = new ChtExternalAppIntentBuilder();
+
+		return builder
+				.setAction(chtExternalApp.getAction())
+				.setCategory(chtExternalApp.getCategory())
+				.setExtras(chtExternalApp.getExtras())
+				.setPackageName(chtExternalApp.getPackageName())
+				.setUri(chtExternalApp.getUri())
+				.setFlags(chtExternalApp.getFlags())
+				.setType(chtExternalApp.getType())
+				.build();
+	}
+
+	static String processResponse(Intent intent, Activity context) {
+		try {
+			if (intent == null || intent.getExtras() == null) {
+				return makeJavaScript(null);
+			}
+
+			JSONObject json = parseBundleToJson(intent.getExtras(), context);
+			return makeJavaScript(json);
+
+		} catch (Exception exception) {
+			error(exception, "ChtExternalAppLauncher :: Problem serialising the intent response");
+			return safeFormat("console.error('Problem serialising the intent response: %s')", exception);
+		}
+	}
+
+	//> PRIVATE
+
+	private static String makeJavaScript(Object response) {
+		String javaScript = "try {" +
+				"const api = window.CHTCore.AndroidApi;" +
+				"if (api && api.resolveCHTExternalAppResponse) {" +
+				"  api.resolveCHTExternalAppResponse(%s);" +
```

Comment:
It looks like this API doesn't exist yet, right? Firstly, we should make sure to version it so we can provide backwards compatibility more easily in future. Secondly I think we should use the same API used in other areas of the Android code, eg: https://github.com/medic/cht-android/blob/53b9646bb69c678b85de304e0a4d684d48e866e1/src/main/java/org/medicmobile/webapp/mobile/MrdtSupport.java#L42

Reply:
The `window.CHTCore.AndroidApi` and `angular.element(document.body).injector().get('AndroidApi')` is the same api coming from webapp, the first one is the new way to invoke (Angular 10+) and the second is the backward compatibility way (AngularJS).
Do you think we should make a new AndroidAPI with version to at least containing this new function? Not sure how often we write these :)

Reply:
It needs to be versioned. The old one was, and looking at the code it looks like this one should be too. If we don't version it then it becomes very difficult to make changes in the future.

Reply:
Okay, I'll try the change in webapp and try to keep the backward compatibility so users don't experience problems if they don't have the latest android app or cht-core

Reply:
Fixed! It already have the versioning :)